File: Backend/app/utils/storage.py
import json
import os
from pathlib import Path
from app.models import UserData

DATA_DIR = Path(__file__).parent.parent.parent / "data"
DATA_DIR.mkdir(exist_ok=True)


def get_user_file(user_id: str) -> Path:
    """Get the path to a user's JSON file."""
    return DATA_DIR / f"{user_id}.json"


def load_user_data(user_id: str) -> UserData:
    """Load user data from JSON file. Create default if not exists."""
    user_file = get_user_file(user_id)
    if user_file.exists():
        with open(user_file, "r") as f:
            data = json.load(f)
            return UserData(**data)
    else:
        # Return default user data
        user = UserData(rfid=user_id, name=f"User {user_id}")
        user.update_timestamp()
        save_user_data(user_id, user)
        return user


def save_user_data(user_id: str, user_data: UserData) -> None:
    """Save user data to JSON file."""
    user_file = get_user_file(user_id)
    user_data.update_timestamp()
    with open(user_file, "w") as f:
        json.dump(user_data.dict(), f, indent=2)


def list_users() -> list:
    """List all user IDs (from JSON files)."""
    return [f.stem for f in DATA_DIR.glob("*.json")]

File: Backend-JS/src/utils/storage.js
// src/utils/storage.js

const fs = require('fs').promises;
const path = require('path');
const UserData = require('../models/User');

const DATA_DIR = path.join(__dirname, '..', '..', 'data');

/**
 * Ensure data directory exists
 */
async function ensureDataDir() {
  try {
    await fs.mkdir(DATA_DIR, { recursive: true });
  } catch (error) {
    console.error('Failed to create data directory:', error);
    throw error;
  }
}

/**
 * Get the path to a user's JSON file
 */
function getUserFilePath(userId) {
  return path.join(DATA_DIR, `${userId}.json`);
}

/**
 * Load user data from JSON file
 * Creates default user data if file doesn't exist
 */
async function loadUserData(userId) {
  try {
    await ensureDataDir();
    const filePath = getUserFilePath(userId);
    
    try {
      const fileContent = await fs.readFile(filePath, 'utf-8');
      const data = JSON.parse(fileContent);
      return new UserData(data.rfid, data.name);
    } catch (error) {
      // File doesn't exist or is invalid, create default
      const newUser = new UserData(userId, `User ${userId}`);
      await saveUserData(userId, newUser);
      return newUser;
    }
  } catch (error) {
    console.error(`Failed to load user data for ${userId}:`, error);
    throw new Error(`Failed to load user data: ${error.message}`);
  }
}

/**
 * Save user data to JSON file
 */
async function saveUserData(userId, userData) {
  try {
    await ensureDataDir();
    userData.updateTimestamp();
    const filePath = getUserFilePath(userId);
    await fs.writeFile(filePath, JSON.stringify(userData.toJSON(), null, 2), 'utf-8');
  } catch (error) {
    console.error(`Failed to save user data for ${userId}:`, error);
    throw new Error(`Failed to save user data: ${error.message}`);
  }
}

/**
 * List all users (from JSON files)
 */
async function listUsers() {
  try {
    await ensureDataDir();
    const files = await fs.readdir(DATA_DIR);
    return files
      .filter(file => file.endsWith('.json'))
      .map(file => path.basename(file, '.json'));
  } catch (error) {
    console.error('Failed to list users:', error);
    throw new Error(`Failed to list users: ${error.message}`);
  }
}

module.exports = {
  loadUserData,
  saveUserData,
  listUsers,
  ensureDataDir
};
